keras_maskrcnn/utils/eval.py

```diff
@@ -0,0 +1,258 @@
+    # Returns
+        A dict mapping class names to mAP scores.
+    """
+    # gather all detections and annotations
+    all_detections, all_masks     = _get_detections(generator, model, score_threshold=score_threshold, max_detections=max_detections, save_path=save_path)
+    all_annotations, all_gt_masks = _get_annotations(generator)
+    average_precisions = {}
+
+    import pickle
+    #all_detections = pickle.load(open('all_detections.pkl', 'rb'))
+    #all_annotations = pickle.load(open('all_annotations.pkl', 'rb'))
+    pickle.dump(all_detections, open('all_detections.pkl', 'wb'))
+    pickle.dump(all_masks, open('all_masks.pkl', 'wb'))
+    pickle.dump(all_annotations, open('all_annotations.pkl', 'wb'))
+    pickle.dump(all_gt_masks, open('all_gt_masks.pkl', 'wb'))
+
+    # process detections and annotations
+    for label in range(generator.num_classes()):
+        false_positives = np.zeros((0,))
+        true_positives  = np.zeros((0,))
+        scores          = np.zeros((0,))
+        num_annotations = 0.0
+
+        for i in range(generator.size()):
+            detections           = all_detections[i][label]
+            masks                = all_masks[i][label]
+            annotations          = all_annotations[i][label]
+            gt_masks             = all_gt_masks[i][label]
+            num_annotations     += annotations.shape[0]
+            detected_annotations = []
+
+            for d, mask in zip(detections, masks):
+                box = d[:4].astype(int)
+                scores = np.append(scores, d[4])
+
+                if annotations.shape[0] == 0:
+                    false_positives = np.append(false_positives, 1)
+                    true_positives  = np.append(true_positives, 0)
+                    continue
+
+                # resize to fit the box
+                mask = cv2.resize(mask, (box[2] - box[0], box[3] - box[1]))
+
+                # binarize the mask
+                mask = (mask > binarize_threshold).astype(np.uint8)
+
+                # place mask in image frame
+                mask_image = np.zeros_like(gt_masks[0])
+                mask_image[box[1]:box[3], box[0]:box[2]] = mask
+                mask = mask_image
+
+                overlaps            = compute_overlap(np.expand_dims(mask, axis=0), gt_masks)
+                assigned_annotation = np.argmax(overlaps, axis=1)
+                max_overlap         = overlaps[0, assigned_annotation]
+
+                if max_overlap >= iou_threshold and assigned_annotation not in detected_annotations:
+                    false_positives = np.append(false_positives, 0)
+                    true_positives  = np.append(true_positives, 1)
+                    detected_annotations.append(assigned_annotation)
+                else:
+                    false_positives = np.append(false_positives, 1)
+                    true_positives  = np.append(true_positives, 0)
+
+        # no annotations -> AP for this class is 0 (is this correct?)
```

Comment:
I am not sure about this, when you do the average of the AP you divide for the total number of classes, shouldn't it be 1?

Reply:
Neither would be correct actually. 0 would indicate something is wrong, 1 would indicate it is perfect, but it is neither..

Reply:
Can we check how other evaluation scripts do?

Reply:
This method was used in py-faster-rcnn, that's the only reference I have :).